Final deliverable

gustavo807 opened the pull request:
-Added unit test for home page, product list, product detail, search results, and shopping cart page

Commit:
Added msw package and babel plugins
Added .babelrc and assetsTransformer files

Commit:
Added msw config files
Added handlres files to mock API endpoints

Commit:
Added custom render with router and redux state to use in the tests

Commit:
Added data-testid to needed components
Added condition to useLatestAPI hook to updated the state only when it's mounted
Fixed typo

Commit:
Unit test files

Commit:
JSON mock files

Commit:
Removed unit test files from the root project

vmcruz commented on src/pages/Cart/__test__/Cart.test.js:
```diff
@@ -0,0 +1,109 @@
+  });
+
+  it("should validate the list of the products is shown when there are items in the cart", () => {
+    const products = initialState.cart.products;
```
This could be rewritten as:

```javascript
const { products } = initialState.cart;
```

vmcruz commented on src/pages/Cart/__test__/Cart.test.js:
```diff
@@ -0,0 +1,109 @@
+    const { format } = formatter;
+
+    for (const product of products) {
+      const price = format(product.data.price);
+      const subtotal = format(product.data.price * product.quantity);
```
Since you are not specifically extracting the price or subtotal, it would be better to rename the variables so that they are descriptive on what data they hold, like: `formattedPrice` and `formattedSubtotal`

vmcruz commented on src/pages/Cart/__test__/Cart.test.js:
```diff
@@ -0,0 +1,109 @@
+  });
+
+  it("should update the total items in the cart when using the quantity selector for a particular product and validate the out of stock", () => {
+    const product = initialState.cart.products[0];
```
This could also be rewritten as:

```javascript
const [firstProduct] = initialState.card.products;
```

vmcruz commented on src/pages/Cart/__test__/Cart.test.js:
```diff
@@ -0,0 +1,109 @@
+
+    fireEvent.click(removeIcon);
+
+    expect(() => screen.getByText(product.data.name)).toThrow(
+      "Unable to find an element"
+    );
```
There's an option here that you could use. Instead of "getting" by text, you should be "querying" by text instead (`queryByText`). The difference between get and query is that query won't `throw` if the element is not found, it will return null. When using "get" you let the testing library that the element is in there, and you want it. The query one is when you are no sure if it is in there or not.

Then use `.not.toBeInTheDocument();` instead of checking for it to throw. Like so:

```javascript
expect(screen.queryByText(product.data.name)).not.toBeInTheDocument();
```

gustavo807 replied:
Thanks, I changed all the toThrow expects for not.toBeInTheDocument, it's more readable.

vmcruz reviewed (commented):
Awesome work in unit testing. I left a couple of comments that can improve the way you test, but good work overall. Also remember that you can check for other things rather than checking if the element exists or not. You can assert the result "toEqual" or "toBe" or if the function "hasBeenCalledTimes".

Commit:
refactor assetsTransformer.js file to export an empty object instead of a function

gustavo807 commented on assetsTransformer.js:
```diff
@@ -0,0 +1,7 @@
+
+module.exports = {
+  process(src, filename, config, options) {
+    return "module.exports = " + JSON.stringify(path.basename(filename)) + ";";
```
I ran into this [Invalid or unexpected token](https://stackoverflow.com/questions/54627028/jest-unexpected-token-when-importing-css) issue when I tried to execute the test.  I added this setting in the package.json to avoid transforming CSS import.
```
"jest": {
    "moduleNameMapper": {
      "\\.(css|less|scss)$": "<rootDir>/assetsTransformer.js"
    },
```
I updated the file to only export an empty object.
`module.exports = {};`

gustavo807 commented on src/hooks/useLatestAPI/index.js:
```diff
@@ -20,15 +21,18 @@ function useLatestAPI() {
       } catch (err) {
-        setApiMetadata({ ref: null, isLoading: false });
-        console.error(err);
+        if (isMounted) {
+          setApiMetadata({ ref: null, isLoading: false });
+          console.error(err);
+        }
```
Yes, I had to add it because in the tests it throws an error.

Commit:
Renamed price and subtotal vars for formatted...
Added array and object destructuring
Changed getByText to queryByText, to use not.toBeInTheDocument instead of toThrow function

Commit:
removed msw worker from the index.js, to use in tests it's already configured in setupTests.js file

gustavo807 commented on src/index.js:
```diff
@@ -1,10 +1,16 @@
+if (process.env.NODE_ENV === "development") {
+  const { worker } = require("./mocks/browser");
+  worker.start();
+}
```
I removed the worker from the here, this config it's only when you want to use it in the browser. For the tests, it's already added in the setupTests.js file

Commit:
Added key to Pagination dots item component
Removed alt default props required property from the Products component
